Use separate parameters to pass bit flags to providers.

slontis opened the pull request:
Changed MAC, digests and ciphers.
NOTE: KDF was already done.

Fixes #12992
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

paulidale commented on doc/man7/provider-cipher.pod:
```diff
@@ -218,13 +218,24 @@ For example AES in CTR mode has a block size of 1 (because it operates like a
 
-=item "flags" (B<OSSL_CIPHER_PARAM_FLAGS>) <unsigned integer>
+=item "aead" (B<OSSL_CIPHER_PARAM_AEAD>) <integer>
+Gets 1 if this is an AEAD cipher algorithm, otherwise it returns 0.
```
Does this need a blank line above?
Also below.

slontis replied:
ta

richsalz commented:
What's the relationship of this PR to #12992 ?

t8m commented:
> What's the relationship of this PR to #12992 ?

It fixes it - as noted in the first commit message.

richsalz commented:
Thanks. I guess it's useful to put the "fixes" entry. in the text as well as the commits.

slontis commented:
Rebased and removed digest cache call that I added (matt added it into the dispatch resolve rather than the fetch).
Hopefully this removes the thread sanitizer issue.

paulidale commented on doc/man7/provider-cipher.pod:
```diff
@@ -218,13 +218,27 @@ For example AES in CTR mode has a block size of 1 (because it operates like a
+
+=item "custom-iv" (B<OSSL_CIPHER_PARAM_CUSTOM_IV>) <integer>
+
+Gets 1 if the cipher algorithm has a custom IV, otherwise it returns 0.
```
_Gets_?  Why not _Returns_?  Alternatively: _returns_?  Why not _gets_?

I don't like gets here, what about: _Has a value of 1 if the cipher algorithm has a custom IV, otherwise it is 0_.

paulidale commented on doc/man7/provider-cipher.pod:
```diff
@@ -218,13 +218,27 @@ For example AES in CTR mode has a block size of 1 (because it operates like a
+
+=item "cts" (B<OSSL_CIPHER_PARAM_CTS>) <integer>
+
+Gets 1 if the cipher algorithm uses ciphertext stealing, otherwise it returns 0.
```
What about: _Has a value of 1 if the cipher algorithm ciphertext stealing, otherwise it is 0_.

paulidale commented on doc/man7/provider-cipher.pod:
```diff
@@ -218,13 +218,27 @@ For example AES in CTR mode has a block size of 1 (because it operates like a
+
+=item "tls-multi" (B<OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK>) <integer>
+
+Gets 1 if the cipher algorithm supports interleaving of crypto blocks, otherwise
```
What about: _Has a value of 1 if the cipher algorithm supports interleaving of crypto blocks, otherwise it is 0_.

slontis replied:
I have stuck with gets as this refers to it being used as a getter.. Has could imply get or set. I have replaced the word returns with gets.

paulidale commented on providers/implementations/include/prov/ciphercommon.h:
```diff
@@ -34,6 +34,15 @@ typedef int (PROV_CIPHER_HW_FN)(PROV_CIPHER_CTX *dat, unsigned char *out,
+#define PROV_CIPHER_FLAG_CUSTOM_IV        0x0002
+#define PROV_CIPHER_FLAG_CTS              0x0004
+#define PROV_CIPHER_FLAG_TLS1_MULTIBLOCK  0x0008
+/* internal flags that are only used within the provider */
```
Capital _I_ at the start of the comment?  Like the previous comment?

paulidale reviewed (approved):
Some wording changes, approved with them updated.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

paulidale commented:
Merged to master
